File: Prob14/Prob14.java
import java.io.*;
import java.util.*;

public class Prob14 {
    public static void main(String[] args) {
        try {
            // set up input handling
            File f = new File("Prob14.in.txt");
            FileReader fr = new FileReader(f);
            BufferedReader br = new BufferedReader(fr);

            // set up hashmap
            HashMap<String, Node> map = new HashMap<String, Node>();

            // set up top level
            Node top = new Node("None");
            map.put("None", top);

            int n = Integer.parseInt(br.readLine());
            for(int i = 0; i < n; i++) {
                String[] currInput = br.readLine().split(",");
                String itemName = currInput[0];
                String parentName = currInput[1];

                // create a new Node for the item
                Node item = new Node(itemName);
                // put the item on the hashmap with the item's name as its key
                map.put(itemName, item);

                // obtain the parent
                Node parent = map.get(parentName);
                // add the child Node to the parent's children
                parent.add(item);
            }

            // print hashmap
            top.print(-1);

        } catch(Exception e) {
            e.printStackTrace();
        }
    }

    public static class Node implements Comparable<Node> {
        String name;
        ArrayList<Node> children;

        // constructor
        public Node(String inputName) {
            name = inputName;
            children = new ArrayList<Node>();
        }

        // adds a child Node to the parent's children
        public void add(Node childNode) {
            children.add(childNode);
        }

        // returns the Node's name
        private String getName() {
            return name;
        }

        // compares the Node's names
        @Override
        public int compareTo(Node otherNode) {
            return name.compareTo(otherNode.getName());
        }

        public void print(int depth) {
            // prints itself if not at the top level
            if(depth >= 0) {
                for(int i = 0; i < depth; i++) {
                    System.out.print("-");
                }
                System.out.println(name);
            }

            // check if node has children
            if(children.size() > 0) {
                // create an array the size of the children ArrayList
                Node[] childArray = new Node[children.size()];
                childArray = children.toArray(childArray);

                // sort childArray alphabetically
                Arrays.sort(childArray);

                // recursively print children
                for(Node child : childArray) {
                    child.print(depth + 1);
                }
            }
        }
    }
}

File: Prob14/Prob14.py
class Node:
    # constructor
    def __init__(self, name):
        self.name = name
        self.children = []

    # used to compare Node objects
    def __eq__(self, other):
        return self.name == other.name
    def __lt__(self, other):
        return self.name < other.name
    def __gt__(self, other):
        return self.name > other.name

    # adds a child Node to children
    def add(self, child):
        self.children.append(child)

    # prints the hierarchy
    def itemPrint(self, depth):
        # prints itself if not at top level
        if depth >= 0:
            for i in range(0, depth):
                print("-", end = "")
            print(self.name)

        # check if Node has children
        if len(self.children) > 0:
            # sort children alphabetically by their name
            self.children.sort()

            # recursively print children
            for child in self.children:
                child.itemPrint(depth + 1)

def main():
    # input handling
    with open("Prob14.in.txt") as f:
        myList = [line.rstrip('\n') for line in f]

        # initialize dictionary
        dictionary = dict()

        # set up top level
        top = Node("None")
        dictionary["None"] = top

        n = int(myList[0])
        for i in range(1, n + 1):
            currInput = myList[i].split(",")
            itemName = currInput[0]
            parentName = currInput[1]

            # create a Node for item
            item = Node(itemName)
            # put item on hashmap
            dictionary[itemName] = item

            # get the parent
            parent = dictionary.get(parentName)
            # add child Node to parent's children
            parent.add(item)

        # print hierarchy
        top.itemPrint(-1)

if __name__ == "__main__": main()
